bgpd: Add command to display EVPN type-5 per-prefix

raja-rajasekar opened the pull request:
Add support for command **show bgp l2vpn evpn route rd <rd> prefix <prefix> [json]**

For Example:
```
e1# sh bgp l2vpn evpn route rd 10.0.0.33:1 prefix 10.1.1.1/32
BGP routing table entry for 10.0.0.33:1:[5]:[0]:[32]:[10.1.1.1]
Paths: (1 available, best #1)
  Advertised to peers:
  10.0.0.1 10.0.0.17
  Route [5]:[0]:[32]:[10.1.1.1] VNI 75100
  1
    120.0.0.1 from 0.0.0.0 (10.0.0.18)
      Origin incomplete, metric 0, valid, sourced, local, best (First path received)
      Extended Community: ET:8 RT:100:75100 Rmac:d2:b1:2b:cb:fd:fc
      Last update: Mon Nov 17 22:01:34 2025

Displayed 1 paths for requested prefix
e1# sh bgp l2vpn evpn route rd 10.0.0.33:1 prefix 10::1/128
BGP routing table entry for 10.0.0.33:1:[5]:[0]:[128]:[10::1]
Paths: (1 available, best #1)
  Advertised to peers:
  10.0.0.1 10.0.0.17
  Route [5]:[0]:[128]:[10::1] VNI 75100
  1
    120.0.0.1 from 0.0.0.0 (10.0.0.18)
      Origin incomplete, metric 0, valid, sourced, local, best (First path received)
      Extended Community: ET:8 RT:100:75100 Rmac:d2:b1:2b:cb:fd:fc
      Last update: Mon Nov 17 22:01:34 2025

Displayed 1 paths for requested prefix

```

vivek-cumulus committed:
bgpd: Add command to display EVPN type-5 per-prefix

Add support for command
show bgp l2vpn evpn route rd <rd> prefix <prefix> [json]

Signed-off-by: Vivek Venkatraman <vivek@cumulusnetworks.com>

Signed-off-by: Rajasekar Raja <rajasekarr@nvidia.com>

raja-rajasekar committed:
tests: Add test for evpn route rd prefix show cmd

Add test for evpn route rd prefix show cmd

Signed-off-by: Rajasekar Raja <rajasekarr@nvidia.com>

mjstapp commented:
Please include an example of the proposed output in the PR description? just saying "support" isn't very ... descriptive.

ton31337 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -5107,6 +5163,70 @@ DEFUN(show_bgp_l2vpn_evpn_route_rd_macip,
+/*
+ * Display global EVPN routing table for specific RD and prefix (type-5).
+ */
+DEFUN(show_bgp_l2vpn_evpn_route_rd_prefix,
```
Please use DEFPY.

raja-rajasekar replied:
Done

ton31337 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -5107,6 +5163,70 @@ DEFUN(show_bgp_l2vpn_evpn_route_rd_macip,
+      "EVPN route information\n"
+      "Route Distinguisher\n"
+      "ASN:XX or A.B.C.D:XX\n"
+      "prefix\n"
```
Can we use capitalized Prefix?

raja-rajasekar replied:
Done

ton31337 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -5107,6 +5163,70 @@ DEFUN(show_bgp_l2vpn_evpn_route_rd_macip,
+		json = json_object_new_object();
+
+	/* get the prd */
+	if (argv_find(argv, argc, "rd", &idx)) {
```
And once you switch to DEFPY, don't use argv_find().

ton31337 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -5107,6 +5163,70 @@ DEFUN(show_bgp_l2vpn_evpn_route_rd_macip,
+	}
+
+	/* get the prefix */
+	if (argv_find(argv, argc, "prefix", &idx)) {
```
Ditto.

ton31337 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -6117,6 +6237,20 @@ ALIAS_HIDDEN(
+	"EVPN route information\n"
+	"Route Distinguisher\n"
+	"ASN:XX or A.B.C.D:XX\n"
+	"prefix\n"
```
Ditto.

raja-rajasekar replied:
Done

ton31337 reviewed (changes requested):
Documentation is missing.

raja-rajasekar committed:
doc: Add show command for EVPN prefix by RD

Document the new show bgp l2vpn evpn route rd prefix command
for displaying EVPN Type 5 routes by route distinguisher and prefix.

Signed-off-by: Rajasekar Raja <rajasekarr@nvidia.com>

raja-rajasekar commented:
ci:rerun

chiragshah6 commented:
https://github.com/FRRouting/frr/pull/21843/

ton31337 commented:
@chiragshah6 to close this one?
